### DISCUSSION TITLE
NHibernate 5.3.x InternalVisibleTo

### DISCUSSION THREAD

**Discussion by TomHriv** [nhibernate/nhibernate-core#2754] [Q&A]
Hello,

I would like to ask you if you know how to use .Net InternalVisibleTo with NHibernate 5.3.x

I need to upgrade project using NHibernate 3.1.x to NHibernate 5.3.x. I have already made a lot of changes in the code, and the project is partly working well, but I need to find a workable solution for the InternalVisibleTo. Because the project is composition of several assemblies and some hbm.xml files and related classes has an internal modifier. I need the internal class to be visible for the NHibernate also.

Our old project was using [assembly: InternalsVisibleTo( "DynamicAssemblyProxyGen, PublicKey=002400.....  this caused that the internal classes were visible for a proxy factory. But this dynamical assembly seems not to be supported in the version 5.3.x. 

I decided to use StaticProxyFactory with NHibernate 5.3.x. Could you please let me know how to amend InternalVisibleTo command?

Thank you very much for help.

**Comment by fredericDelaporte** [comment]
I have not checked what was the state of the dynamic proxy code back in NHibernate v3.1.x, but it seems the assembly naming has changed since then, be it with the now obsolete dynamic proxy or with the current static proxy.
The current naming is done here for the static proxy:
https://github.com/nhibernate/nhibernate-core/blob/07ecb9f25534d3894dfba6a3a8e8f22932bd06cf/src/NHibernate/Proxy/NHibernateProxyBuilder.cs#L50-L51
So usually, if the proxied entity class is named `Entity`, the proxy assembly will be named `EntityProxyAssembly`.

For the dynamic proxy, it is here, and that is doing the same thing, just with an older syntax:
https://github.com/nhibernate/nhibernate-core/blob/07ecb9f25534d3894dfba6a3a8e8f22932bd06cf/src/NHibernate/Proxy/DynamicProxy/ProxyFactory.cs#L81-L82

Will it be enough for your case? I have no idea. I do not think relying of `InternalsVisibleTo` is a good solution, even if it may work. Of course I do not know why you need to have these classes internal, but if you could switch them to public, or map them through public interfaces, that would be more reliable.

**Reply by TomHriv** [reply]
Hello, thank you for answer. The switching internal classes to public is my plan B.

**Comment by bahusoid** [comment]
Internal entity classes should be supported in 5.4 (https://github.com/nhibernate/nhibernate-core/pull/2568). So as an option you can try to use 5.4 development builds (afaik it's in pretty stable state right now): https://www.myget.org/feed/nhibernate/package/nuget/NHibernate

Note: It doesn't work with .net standard.  Use .NET Core/.NET Framework NHibernate versions to make it work.
